mlflow/types/schema.py

```diff
@@ -406,3 +458,280 @@ def __eq__(self, other) -> bool:
+        """Default value of the parameter."""
+        return self._default
+
+    @property
+    def shape(self) -> Optional[tuple]:
+        """
+        The parameter shape.
+        If shape is None, the parameter is a scalar.
+        """
+        return self._shape
+
+    class ParamSpecTypedDict(TypedDict):
+        name: str
+        type: str
+        default: Union[DataType, List[DataType], None]
+        shape: Optional[Tuple[int, ...]]
+
+    def to_dict(self) -> ParamSpecTypedDict:
+        if self.shape is None:
+            if self.type.name == "datetime":
+                default_value = np.datetime_as_string(self.default)
+            else:
+                default_value = self.type.to_python()(self.default)
+        elif self.shape == (-1,):
+            if self.type.name == "datetime":
+                default_value = list(map(np.datetime_as_string, self.default))
+            else:
+                default_value = list(map(self.type.to_python(), self.default))
+        return {
+            "name": self.name,
+            "type": self.type.name,
+            "default": default_value,
+            "shape": self.shape,
+        }
+
+    def __eq__(self, other) -> bool:
+        if isinstance(other, ParamSpec):
+            return (
+                self.name == other.name
+                and self.type == other.type
+                and self.default == other.default
+                and self.shape == other.shape
+            )
+        return False
+
+    def __repr__(self) -> str:
+        shape = f" (shape: {self.shape})" if self.shape is not None else ""
+        return f"{self.name!r}: {self.type!r} (default: {self.default}){shape}"
+
+    @classmethod
+    def from_json_dict(cls, **kwargs):
+        """
+        Deserialize from a json loaded dictionary.
+        The dictionary is expected to contain `name`, `type` and `default` keys.
+        """
+        if not {"name", "type", "default"} <= set(kwargs.keys()):
+            raise MlflowException(
+                "Missing keys in ParamSpec JSON. Expected to find keys `name`, `type` and `default`"
+            )
+        return cls(
+            name=str(kwargs["name"]),
+            type=DataType[kwargs["type"]],
+            default=kwargs["default"],
+            shape=kwargs.get("shape"),
+        )
+
+
+@experimental
+class ParamSchema:
+    """
+    Specification of parameters applicable to the model.
+    ParamSchema is represented as a list of :py:class:`ParamSpec`.
+    """
+
+    def __init__(self, params: List[ParamSpec]):
+        if not all(isinstance(x, ParamSpec) for x in params):
+            raise MlflowException(f"ParamSchema inputs only accept {ParamSchema.__class__}")
+        if duplicates := self._find_duplicates(params):
+            raise MlflowException(f"Duplicated parameters found in schema: {duplicates}")
+        self._params = params
+
+    @staticmethod
+    def _find_duplicates(params: List[ParamSpec]) -> List[str]:
```

Comment:
Can we add a test that validates this behavior? I think it's the only error handling that's not covered in the `test_pyfunc_schema_enforcement` suite below

Reply:
Ok!